### ISSUE TITLE
About parameter base_estimator assignment in SPISEClassifier class

### ISSUE THREAD

**Issue by xuzhou666** [CUHKSZ-NING/SPISEClassifier#1]
When I assign only one value to base_estimator, for example ‘base_estimator = RandomForestClassifier()’, the code runs fine. When I want to use multiple base classifiers and assign multiple values to base_estimator, what type of data do I need to use? Tried types like list, dictionary, etc. e.g. ‘base_estimator = {svm.SVC(probability=True), RandomForestClassifier()}’, after that an error is reported: ''dict' object has no attribute 'fit''. 
Or is setting multiple base classifiers not changing the value of base_estimator?
If possible hope to get an answer, thanks!

**Comment by CUHKSZ-NING** [comment]
You may define a voting classifier class (that out puts the voting of SVC and RF) with fit(), predict(), and predict_proba() functions. Direct input of a dict is currently not supported.
Sincerely,
Zhihan Ning.

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: xuzhou666 ***@***.***>
Sent: Sunday, July 16, 2023 11:10:25 AM
To: CUHKSZ-NING/SPISEClassifier ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [CUHKSZ-NING/SPISEClassifier] About parameter base_estimator assignment in SPISEClassifier class (Issue #1)


When I assign only one value to base_estimator, for example ‘base_estimator = RandomForestClassifier()’, the code runs fine. When I want to use multiple base classifiers and assign multiple values to base_estimator, what type of data do I need to use? Tried types like list, dictionary, etc. e.g. ‘base_estimator = {svm.SVC(probability=True), RandomForestClassifier()}’, after that an error is reported: ''dict' object has no attribute 'fit''.
Or is setting multiple base classifiers not changing the value of base_estimator?
If possible hope to get an answer, thanks!

―

**Comment by xuzhou666** [comment]
> 

I'm very sorry I don't quite understand what you mean. isn't the voting classifier class already defined as the VotingEstimator class? In each iteration, self.estimators[i] is given the attributes of VotingEstimator? Do you mean that only one base classifier can be given at a time? For example, the 7 base classifiers mentioned as being used in the paper, is one of them chosen each time to complete the training prediction, and this is repeated 7 times? So in the last formula of the pseudo-code of Algorithm 1:$F(\cdot)=\frac{1}{k} \sum_{i=1}^k f_i(\cdot)$, integrating the results for i base classifiers, is the base classifier f(-) the same?

**Comment by xuzhou666** [comment]
I tried changing the code here by iteratively assigning the dictionary key that holds the base classifier to the VotingEstimator class as the base classifier, with the input dictionary formatted as shown in the figure, at which point the code worked correctly, but did not rigorously verify the accuracy of the results
![image](https://github.com/CUHKSZ-NING/SPISEClassifier/assets/76723484/2b6caa70-a17c-4111-bc38-48df74eb0dd5)
![image](https://github.com/CUHKSZ-NING/SPISEClassifier/assets/76723484/c5f019b1-e0ca-40da-8468-495205a3d8ef)

**Comment by CUHKSZ-NING** [comment]
Basically, SPISE is designed for constructing a number of balanced subsets for imbalanced classification. A base classifier is required to fit these subsets for prediction. In the paper, the results of seven SEPARATE classifiers (e.g., SVM, LightGBM) are provided. In fact, the combination of several classifiers, say SVM and RF, as an ensemble is not the key of this paper and is thus not tested.

**Comment by xuzhou666** [comment]
Okay, I understand then, thanks a lot for the answer!

**Comment by xuzhou666** [comment]
Okay, I understand then, thanks a lot for the answer!
